tests/run-pass/concurrency/locks.rs

```diff
@@ -0,0 +1,29 @@
+//! This test just calls the relevant APIs to check if Miri crashes.
```

Comment:
I doubt this test will pass on Windows.

Did you test macOS? You can do that on a linux box with `--target`/`MIRI_TEST_TARGET`.

Reply:
> Did you test macOS? You can do that on a linux box with --target/MIRI_TEST_TARGET.
Thanks! I have just tried. The test fails on [set_global_dtor](https://github.com/rust-lang/miri/blob/a1c9ad8e1b239727ec0b86440cdc7a662d7550e5/src/shims/tls.rs#L105). Should be there a `global_dtor` per thread? If not, do you know what mechanism is used to ensure that there exists only one / could you point me to the documentation you used to implement the current version?

Reply:
I don't know macOS APIs so I don't know if they are per-thread or not -- but I assume "global" to mean "not per thread". What I wrote is based on looking at l[ibstd code that calls this shim](https://github.com/rust-lang/rust/blob/534a41a32952d36ec73656357777ebbea707aeb4/src/libstd/sys/unix/fast_thread_local.rs#L61) and the [actual source of this function on macOS](https://github.com/opensource-apple/dyld/blob/195030646877261f0c8c7ad8b001f52d6a26f514/src/threadLocalVariables.c#L387).

Reply:
Thanks for pointing to the relevant code! This is very helpful. The call to `_tlv_atexit` is clearly [guarded by a thread local](https://github.com/rust-lang/rust/blob/534a41a32952d36ec73656357777ebbea707aeb4/src/libstd/sys/unix/fast_thread_local.rs#L58-L60), so I guess it is per thread...

Reply:
After changing the `global_dtor` to be per thread, the tests on Mac OS pass.

Reply:
> Thanks for pointing to the relevant code! This is very helpful. The call to _tlv_atexit is clearly guarded by a thread local, so I guess it is per thread...

That could just mean Rust wants a global destructor per thread.

However, the macOS source code says
```
// NOTE: this does not need locks because it only operates on current thread data
```
so that quite clearly looks like it all acts per-thread. Could you add a link to that in a comment in the code?